internal/test/deployment/pmm.go

```diff
@@ -0,0 +1,387 @@
+		},
+	})
+	if err != nil {
+		return err
+	}
+	if len(netresp.Warning) > 0 {
+		return errors.New("got warnings during network creation:" + netresp.Warning)
+	}
+
+	pmm.Log("Creating PMM server")
+	if err := pmm.CreatePMMServer(ctx, dockerCli, netresp.ID); err != nil {
+		return errors.Wrap(err, "failed to create pmm server")
+	}
+
+	pmm.Log("Creating PMM client")
+	if err := pmm.CreatePMMClient(ctx, dockerCli, netresp.ID); err != nil {
+		return errors.Wrap(err, "failed to create pmm client")
+	}
+
+	pmm.Log("Creating mongo")
+	if err := pmm.CreateMongo(ctx, dockerCli, netresp.ID); err != nil {
+		return errors.Wrap(err, "failed to create mongo container")
+	}
+
+	pmm.Log("Waiting for mongo to be ready")
+	err = doUntilSuccess(60*time.Second, func() error {
+		return pmm.PingMongo(ctx)
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed to connect to mongo")
+	}
+
+	pmm.Log("Adding mongo to PMM")
+	if err := doUntilSuccess(60*time.Second, func() error {
+		return pmm.Exec(ctx, pmm.ClientContainerName(),
+			"pmm-admin", "add", "mongodb",
+			"--username", "admin",
+			"--password", "admin",
+			"mongo",
+			pmm.MongoContainerName()+":27017")
+	}); err != nil {
+		return errors.Wrap(err, "failed to add mongo to PMM")
+	}
+
+	if err := doUntilSuccess(60*time.Second, func() error {
+		return pmm.PingClickhouse(ctx)
+	}); err != nil {
+		return errors.Wrap(err, "failed to ping clickhouse")
+	}
+
+	return nil
+}
+
+func (pmm *PMM) Restart(ctx context.Context) error {
+	dockerCli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return errors.Wrap(err, "failed to create docker client")
+	}
+	defer dockerCli.Close() //nolint:errcheck
+
+	if err := dockerCli.ContainerRestart(ctx, pmm.pmmServerContainerID, container.StopOptions{
+		Timeout: nil, // 10 seconds
+	}); err != nil {
+		return errors.Wrap(err, "failed to restart pmm server")
+	}
+	if err := pmm.SetServerPublishedPorts(ctx, dockerCli); err != nil {
+		return errors.Wrap(err, "failed to set server published ports")
+	}
+
+	if err := getUntilOk(pmm.PMMURL()+"/v1/version", time.Second*30); err != nil {
+		return errors.Wrap(err, "failed to ping PMM")
+	}
+	return nil
+}
+
+func (pmm *PMM) Destroy(ctx context.Context) {
+	pmm.Log("Destroying deployment")
+	if err := destroy(ctx, filters.NewArgs(filters.Arg("label", PerconaLabel+"="+pmm.testName))); err != nil {
+		pmm.Log(err)
+		pmm.t.FailNow()
+	}
+}
+
+func getUntilOk(url string, timeout time.Duration) error {
+	return doUntilSuccess(timeout, func() error {
+		resp, err := http.Get(url) //nolint:gosec,noctx
+		if err != nil {
+			return err
+		}
+		defer resp.Body.Close() //nolint:errcheck
+		if resp.StatusCode == http.StatusOK {
+			return nil
+		}
+		return errors.New("not ok")
+	})
+}
+
+func doUntilSuccess(timeout time.Duration, f func() error) error {
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+	timeoutTimer := time.NewTimer(timeout)
```

Comment:
Probably context with timeout will fit better here. I think this func can just accept ctx as first parameter, and context with timeout should be created on the invocation side.

Reply:
https://github.com/percona/pmm-dump/pull/99/commits/24e8b9804689c5470b3ca023822a3214eff3d7da

Reply:
@pooknull  I mean that context with timeout can be set outside function, like this:
```
ctx, cancel := context.WithTimeout(ctx, timeout)
defer cancel()

err := doUntilSuccess(ctx, f)
```

This is more idiomatic way to apply timeout restictions

Reply:
@artemgavrilov, you wrote this is LGTM, but @pooknull implemented the suggested change. Please either resolve this conversation or add a comment about you are waiting for the better fix.

Reply:
@artemgavrilov My idea was to avoid the possible error of passing ctx without timeout. But it's not critical, so let's make it idiomatic: https://github.com/percona/pmm-dump/pull/99/commits/6d604eb76e223919d3a014e1fc7cbf52c2191ebb